### ISSUE TITLE
MRView: tractography display issues on ATI hardware

### ISSUE THREAD

**Issue by jdtournier** [MRtrix3/mrtrix3#271]
I just came across this issue on someone's system running with a fairly basic ATI HD 7500 card (Ubuntu 14.04, 64 bit, fglrx proprietary drivers). It seems it really can't handle the new geometry shader pipeline, but I'm not sure whether it's an issue with how we do things or whether it's a bug in their drivers (in my experience, the ATI drivers tend to be more picky than the Nvidia ones, but nonetheless standard-compliant). My guess is there is probably something non-compliant about what we're doing that's causing these issues. 

Basically, only a fraction of the streamlines get displayed. Those that do don't look very nice, and their ends come out black for some reason. The framerate is also really bad. Obviously, everything runs fine when running a previous version, before the geometry shader branch was merged in... 

What we do about this will probably depend on whether we can identify anything about our implementation that might be causing these issues. I'll try peppering the code with OpenGL debugging statements, see whether that throws up anything...

**Comment by jdtournier** [comment]
OK, just replicated on NVIDIA hardware too... At least the black ends on streamlines - performance seems fine, and I don't seem to have any missing streamlines. But it does point to some issue somewhere in the tractography tool. I'll investigate if/when I get the chance. In the meantime, I'd be interested to hear whether anyone can replicate. And of course, if anyone is at a loss for something to do, feel free to investigate too...

**Comment by thijsdhollander** [comment]
Are the black ends only showing when using partial transparency of the tracks, while combined with a volume render?  I also get that, but it seems to be how the tractogram renderer currently deals with transparency (making it fully transparent, still fills its silhouette with the background colour)... In that case, the black is actually the background colour (which can now be set to any other colour).  Don't know if it would give any useful clues, but it might be worth checking if it's "just black", or rather the background colour, by changing the latter to something else.  Maybe the ends are for some reason not rendered, and thus the background colour shows instead...? :confused: 

In some cases, if I set the trackwidth to the smallest possible value (it even defaults to that in some/all cases, I think), I also get "gaps" along the track.

**Comment by jdtournier** [comment]
@thijsdhollander : see #278 for some examples of the issue - the last screenshot on my comment is pretty much what we see on the ATI hardware - with the exception that on top of that, most of the tracks don't show up at all.

**Comment by thijsdhollander** [comment]
Ok, so specifically on these black ends: that's definitely not what I'm getting or have spotted anywhere else.  What I was talking about, was more related to how (specifically) the volume render interacts with tracks.  I'll be on the lookout for the specific symptoms in those screenshots; not expecting to get them since I'm using neither ATI hardware nor qt 5.4.  But given that they seem to also partially apply to scenarios where one of those "requirements" is not present... they might indeed be more general bugs that just apply to very specific rare circumstances we're not really understanding yet...

**Comment by jdtournier** [comment]
Finally had a look at exactly how the shaders are set up for the streamlines render. It seems things are done in quite a different way from what I had in mind, and this may be causing some of the issues here. 

What's currently done is to pass each vertex _and the next_ to the vertex shader, which uses this information to compute the segment tangent. All the geometry shader does then is take this information and amplify the number of vertices. This will probably cause issues at track ends where you'd expect to find a triplet of NaNs (which may explain the slowdown on ATI hardware, depending on how they handle NaNs, it can often be a pretty slow process on the CPU). It would also potentially explain why some of the tangent vectors are corrupt.

What I thought was happening was that the vertex shader would operate per vertex and do nothing more than set the `gl_Position`, and maybe also compute whether the vertex is visible. The computation of the tangent and associated vectors would be done within the geometry shader, assuming an input layout of `lines_adjacency`. That would allow the shader to catch NaN's and process these vertices appropriately, and also compute everything it needs to from there. I notice in the git history that this is in fact how things were done before - why was this changed...? Was performance affected in some way? Or was this is response to my inane ramblings when trying to avoid using a geometry shader in the first place... (issue #91)?

**Comment by rtabbara** [comment]
@jdtournier
<blockquote>
This will probably cause issues at track ends where you'd expect to find a triplet of NaNs (which may explain the slowdown on ATI hardware, depending on how they handle NaNs
</blockquote>

Ah, this is probably the issue then. Not sure whether it's the case that Nvidia cards more gracefully handle NaNs, but given that everything rendered nicely, I assumed that track data was tightly packed and not delimited by a NaN point. I think adjusting the range of vertices/vertex attributes to filter out any NaNs when loading onto the GPU should sort this out. Sorry, I had a feeling this might be the issue but I was feeling unwell so didn't get a chance to look at this today.

<blockquote>
- why was this changed...? Was performance affected in some way? Or was this is response to my inane ramblings when trying to avoid using a geometry shader in the first place...
</blockquote>

Yep, the optimisation suggestions I came across were to make the geometry shader code as lightweight as possible and move more computation down the pipeline to the vertex shader. The motivation is that while the geom. shader is taking its sweet time, at least something useful can be done while waiting.

**Comment by jdtournier** [comment]
I had a go at modifying the approach taken in the tractography tool to match what I'd originally had in mind - check out #281, see what you think. This is a first crack at it, there's still a few things to do, but with my limited testing, on my setup the framerate seemed much better with a large tractogram than using the current master. Not completely sure why, both approaches are pretty similar, really.

Just to give an idea of what I've done: 
- the vertex shader is now very lightweight - computes position, the exclusion flag, and the slab coordinate if cropping to slab. It only uses one vertex attribute for the main vertex position (although that might change - see later).
- the geometry shader now handles computing the tangent, colour, discarding the primitive if all exclusion flags are set, etc. In this case, I only compute 4 vertices for a single quad (rather than 6), which overlap at each end by the streamline width. The idea will eventually be to get the fragment shader to round off the corners, if that proves necessary.
- the fragment shader does pretty much what it did before: crop to slab, colouring & lighting.

Things left to do are to get per-vertex colour when colouring by direction, and interpolating that over the segment (currently each segment is coloured flat). This would be done in the vertex shader using two shifted vertex attributes - we know it worked well in the previous implementation, so I'm hoping it'll be nice and quick here too. Other thing to do would be trying to round off the corners by discarding in the fragment shader - although that might hurt performance without improving the appearance all that much, not sure, we'll have to see...

**Comment by jdtournier** [comment]
OK, tested the latest version on ATI - works fine. Only remaining issue on ATI is #285, which is likely unrelated... Anyway, I'll close this one off.
